byzcoin/trie/trie.go

```diff
@@ -0,0 +1,648 @@
+		}
+		if !bytes.Equal(key, node.Key) {
+			// key doesn't exist, nothing to delete
+			return nil, nil
+		}
+		if err := b.Delete(node.hash(t.nonce)); err != nil {
+			return nil, err
+		}
+		empty := newEmptyNode(node.Prefix)
+		emptyBuf, err := empty.encode()
+		if err != nil {
+			return nil, err
+		}
+		if err := b.Put(empty.hash(t.nonce), emptyBuf); err != nil {
+			return nil, err
+		}
+		return empty.hash(t.nonce), nil
+	case typeInterior:
+		node, err := decodeInteriorNode(nodeVal)
+		if err != nil {
+			return nil, err
+		}
+		// update this interior node
+		if bits[depth] {
+			// look left
+			res, err := t.del(depth+1, node.Left, bits, key, b)
+			if err != nil {
+				return nil, err
+			}
+			if res == nil {
+				// not found, so do nothing
+				return nil, nil
+			}
+			// delete the old interior node
+			if err := b.Delete(node.hash()); err != nil {
+				return nil, err
+			}
+			// update this interior node
+			node.Left = res
+			nodeBuf, err := node.encode()
+			if err != nil {
+				return nil, err
+			}
+			return node.hash(), b.Put(node.hash(), nodeBuf)
+		}
+		// look right
+		res, err := t.del(depth+1, node.Right, bits, key, b)
+		if err != nil {
+			return nil, err
+		}
+		if res == nil {
+			// not found, so do nothing
+			return nil, nil
+		}
+		// delete the old interior node
+		if err := b.Delete(node.hash()); err != nil {
+			return nil, err
+		}
+		// update this interior node
+		node.Right = res
+		nodeBuf, err := node.encode()
+		if err != nil {
+			return nil, err
+		}
+		return node.hash(), b.Put(node.hash(), nodeBuf)
+	}
+	return nil, errors.New("invalid node type")
+}
+
+// IsValid checks whether the trie is valid.
+func (t *Trie) IsValid() error {
+	p := countNodeProcessor{}
+	err := t.db.View(func(b bucket) error {
+		rootKey := t.getRoot(b)
+		if rootKey == nil {
+			return errors.New("no root key")
+		}
+		return t.dfs(&p, rootKey, b)
+	})
+	if err != nil {
+		return err
+	}
+
+	// We can get proof for all the leaves.
+	for _, leave := range p.leaves {
+		proof, err := t.GetProof(leave.Key)
+		if err != nil {
+			return err
+		}
+
+		proof.noHashKey = t.noHashKey
+
+		ok, err := proof.Exists(leave.Key)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return errors.New("got absence proof")
+		}
+	}
+
+	// Check that we have no dangling nodes.
+	var total int
+	err = t.db.View(func(b bucket) error {
+		return b.ForEach(func(k, v []byte) error {
+			total++
+			return nil
+		})
+	})
+	if err != nil {
+		return err
+	}
+	if total != p.total+2 {
+		// plus 2 because there are two well-known keys
+		return errors.New("dangling nodes")
+	}
+	return nil
+}
+
+// getRaw gets the value, it returns nil if the value does not exist.
+func (t *Trie) getRaw(key []byte) ([]byte, error) {
+	var val []byte
+	err := t.db.View(func(b bucket) error {
+		buf := b.Get(key)
+		val = clone(buf)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+func (t *Trie) binSlice(buf []byte) []bool {
```

Comment:
I'm unclear why this code to change []byte into []bool exists; it seems like the only thing we do with the resulting []bool is to do lookups into it, i.e. `if bits[i] ...etc...`. It seems to me like checking if bit #n in []byte x is set should not require a []bool.

Reply:
I initially thought it was possible to remove the `binSlice` functions, but then I realised they're not only used for iteration, it's also important to store and use `[]bool` that are not on "byte-boundaries". that's because the way we make new nodes is when their prefix becomes different, so two nodes may have the same prefix for 5 bits, then the 6th one is different. in that case we have to store a `[]bool` of length 6 in the leaf node, see `extendLeaf`. it wouldn't be possible to do with just a `[]byte` without additional information. bool in go is one byte (https://play.golang.org/p/jkfkv5LJSqU), hopefully that shouldn't be too bad on the storage